func_sim/func_memory/func_memory.cpp

```diff
@@ -11,39 +11,148 @@
+                    memset( Memory[ first][ second], 0, 1000 * sizeof( uint8));
+                }
+            }
+        if ( offset_size_global < 32)	
+            Memory[ first][ second][ third] = ( uint8) * ( sections_array[ i].content + j);
+        }
+        i++;
+    }
 }
 
 FuncMemory::~FuncMemory()
 {
-    // put your code here
+    for ( int i = 0; i < addr_count; i++)
+        for ( int j = 0; j < page_num_count; j++)
+            if ( Memory[ i] != NULL)
+                if ( Memory[ i][ j] != NULL)
+                    delete[] Memory[i][j];
+    for ( int i = 0; i < addr_count; i++)
+        if ( Memory[ i] != NULL)
+            delete[] Memory[ i];
+    if ( Memory != NULL)
+        delete[] Memory;
 }
 
 uint64 FuncMemory::startPC() const
 {
-    // put your code here
+    int i = 0;
+    while ( strcmp( ".text", sections_array[ i].name) != 0)
+        i++;
+    return sections_array[ i].start_addr;
 }
 
 uint64 FuncMemory::read( uint64 addr, unsigned short num_of_bytes) const
 {
-    // put your code here
-
-    return 0;
+    if ( !num_of_bytes)
+        abort();
+    uint64 first, second, third;
+    uint64 returnvalue = 0;
+    for ( int j = num_of_bytes - 1; j >= 0; j--)
+    {
+        FuncMemory::memoryCalculator( &first, &second, &third, addr + j * sizeof( uint8));
+        if ( first >= addr_count 
+          || second >= page_num_count
+          || third >= offset_count)
+            abort();
+        if ( Memory[ first] == NULL) 
+            abort();
+        if ( Memory[ first][ second] == NULL) 
+            abort();
+        returnvalue = 256 * returnvalue + ( uint32) Memory[ first][ second][ third];
+    }
+    return returnvalue;
 }
 
 void FuncMemory::write( uint64 value, uint64 addr, unsigned short num_of_bytes)
 {
-    // put your code here
+    if ( !num_of_bytes)
+        abort();
+    uint64 first, second, third;
+    for ( int j = 0; j < num_of_bytes; j++)
+    {
+        memoryCalculator(&first, &second, &third, addr + j * sizeof(uint8));
+        if (first >= addr_count || 
+            second >= page_num_count || 
+            third >= offset_count || 
+            first < 0 || 
+            second < 0|| 
+            third < 0) 
+                abort();
+        
+        if ( Memory[ first] == NULL)
+        {
+            Memory[ first] = new uint8*[ page_num_count];
+            memset( Memory[ first], 0, page_num_count * sizeof( uint8*));
+        }
+        if (Memory[ first][ second] == NULL) 
+            {
+                Memory[ first][ second] = new uint8[ offset_count];
+                memset( Memory[ first][ second], 0, offset_count * sizeof( uint8));
+            }
+        if (offset_size_global < 32)
+            Memory[ first][ second][ third] = ( uint32)( value % 256);
+            value /= 256;
+    }
 }
 
 string FuncMemory::dump( string indent) const
 {
-    // put your code here
-    return string("ERROR: You need to implement FuncMemory!");
+    return string("Not yet!");
```

Comment:
I recommend you to look at ELF parser *dump()* method if you face difficulties with its realization.